primitives/runtime/src/traits.rs

```diff
@@ -961,12 +961,20 @@ pub trait Extrinsic: Sized {
 	/// The function call.
 	type Call;
 
-	/// The payload we carry for signed extrinsics.
+	/// The type of the address that signed the extrinsic.
 	///
-	/// Usually it will contain a `Signature` and
-	/// may include some additional data that are specific to signed
-	/// extrinsics.
-	type SignaturePayload;
```

Comment:
Instead of breaking most of the code we could just do
```
type SignaturePayload: SignaturePayload;
```

```
trait SignaturePayload {
      type Address: TypeInfo;
      type Signature: TypeInfo;
      type Extra: TypeInfo;
}
```

Reply:
Yep, I like this approach more! Thanks for the feedback!

Reply:
@bkchr What's your opinion on having the `extrinsic.ty` in the metadata after this change (the message James mentioned above)?

We should have enough type info by relying on:
- address_ty
- call_ty
- signature_ty
- extra_ty

The `extrinsic.ty` has a manual implementation of `TypeInfo` that adds the above ty as generic types, but since we explicitly expose that info with this PR, I believe we should remove the `extrinsic.ty`.

That is unless we'll break other things

Reply:
I have no idea if someone is using this. However, frankly speaking by only having the types you listed you don't know the encoding of the `Extrinsic` type. While currently it may also already not being correct as it probably doesn't include the `Compact<u32>` length in from of the extrinsinc encoding?

Reply:
Good idea! I'll add the `TypeDefCompact<u32>` type to the extrinsic metadata to provide even better decoding abilities to end users. Considering this, I'll remove the `extrinsic_ty` from the metadata, unless encountering some edge-cases from UX teams.

The type only contained the `address`,  `call`, `signature` and `extra` on its type info definition, and users should be able to obtain those details (and more) directly.

Reply:
iirc the extrinsic type was quite useless anyway before; it mostly just said "hey we want to encode the extrinsic to some vec of bytes". It doesn't provide any useful information on how those bytes need to be formed etc.

There are already a bunch of things we must just assume already about encoding extrinsics, ie the order of the signature stuff, how the not-really-scale Option for the signature is encoded (ie it has a version number `&&`'d to it) etc. I've always assumed this format information to be encoded in the tx version (4 atm); if a transaction has this version, it must be encoded in this way. 

The metadata, in my mind, only needs to provide the information that might vary from chain to chain (eg the actual additional params etc). This tx version can encode the actual format that this stuff is turned into an extrinsic.

All of this to say, what would adding a  `TypeDefCompact<u32>` to the metadata help with? I'd assume it to be part of the V4 extrinsic format I guess, but perhaps I'm wrong on this :)

Reply:
> There are already a bunch of things we must just assume already about encoding extrinsics, ie the order of the signature stuff, how the not-really-scale Option for the signature is encoded (ie it has a version number `&&`'d to it) etc. I've always assumed this format information to be encoded in the tx version (4 atm); if a transaction has this version, it must be encoded in this way.

Good point. 

> The metadata, in my mind, only needs to provide the information that might vary from chain to chain (eg the actual additional params etc). This tx version can encode the actual format that this stuff is turned into an extrinsic.

Just because every body is currently using our `UncheckedExtrinsic` doesn't mean that this will always be the case. However, it makes sense that this is a type that we can not depict in as type metadata.